composer/callbacks/speed_monitor.py

```diff
@@ -63,106 +63,106 @@ class SpeedMonitor(Callback):
+            self.epoch_start_wct = self._loaded_state["epoch_start_wct"]
+            self.batch_start_num_samples = self._loaded_state["batch_start_num_samples"]
+            self.batch_start_wct = self._loaded_state["batch_start_wct"]
+            self.rolling_batch_times_and_num_samples = deque(
+                [x for x in self._loaded_state["rolling_batch_times_and_num_samples"]],
+                maxlen=self.window_size,
+            )
+            self.running_wct = self._loaded_state["running_wct"]
+            self.running_num_samples = self._loaded_state["running_num_samples"]
+            self.total_eval_wct = self._loaded_state["total_eval_wct"]
+            self._loaded_state = None
 
-    def batch_start(self, state: State, logger: Logger) -> None:
+    def epoch_start(self, state: State, logger: Logger):
         del logger  # unused
         self._load_state()
-        self.batch_start_num_samples = state.timestamp.sample
+        self.rolling_batch_times_and_num_samples.clear()
+        self.epoch_start_wct = state.timestamp.total_wct
+        self.epoch_start_num_samples = state.timestamp.sample
 
-    def epoch_start(self, state: State, logger: Logger):
-        del state, logger  # unused
+    def batch_start(self, state: State, logger: Logger) -> None:
+        del logger  # unused
         self._load_state()
-        self.epoch_start_time = time.time()
-        self.batch_end_times.clear()
-        self.batch_num_samples.clear()
-        self.train_examples_per_epoch = 0
+        self.batch_start_wct = state.timestamp.total_wct
+        self.batch_start_num_samples = state.timestamp.sample
 
     def batch_end(self, state: State, logger: Logger):
-        self.batch_end_times.append(time.time())
         new_num_samples = state.timestamp.sample
-        assert self.batch_start_num_samples is not None, "self.batch_start_num_samples should have been set on Event.BATCH_START"
-        batch_num_samples = int(new_num_samples - self.batch_start_num_samples)
-        self.batch_num_samples.append(batch_num_samples)
-        self.train_examples_per_epoch += batch_num_samples
-        if len(self.batch_end_times) == self.window_size + 1:
-            throughput = sum(self.batch_num_samples) / (self.batch_end_times[-1] - self.batch_end_times[0])
+        batch_num_samples = new_num_samples - self.batch_start_num_samples
+        self.running_num_samples += batch_num_samples
+        batch_wct = state.timestamp.total_wct - self.batch_start_wct
+        self.running_wct += batch_wct
+
+        if len(self.rolling_batch_times_and_num_samples) == self.window_size:
+            # the buffer is full. subtract out the oldest num samples and wct from the running tallies, before the new entry is added
+            oldest_wct, oldest_num_samples = self.rolling_batch_times_and_num_samples.popleft()
+            self.running_num_samples -= oldest_num_samples
+            self.running_wct -= oldest_wct
```

Comment:
how much does this optimization save us? I think we are saving a .sum() of a few numbers here, if I understand this correctly. Maybe it's not worth having to keep extra variables and this logic here, and just let deque do its thing?